### ISSUE TITLE
Member Base Substitution Has a Bug

### ISSUE THREAD

**Issue by lenary** [Microsoft/checkedc-clang#342]
I think this is an issue, but given lack of a testcase i cannot prove it will ICE.

in SemaBounds.cpp `ConcretizeMemberBounds` class, there is the following code:
```cpp
    ExprResult TransformDeclRefExpr(DeclRefExpr *E) {
      if (FieldDecl *FD = dyn_cast<FieldDecl>(E->getDecl())) {
        if (Base->isRValue() && !IsArrow)
          // For now, return nothing if we see an rvalue base.
          return ExprResult();
      // ... more 
```

We shouldn't return an ExprResult() when we don't know how deep in the expression we are. It will not short-circuit and stop substituting, it will return a `nullptr` back up to the next level up in place of the `Expr*` it could be returning. If the next level up doesn't want that (say it's a BinaryOperator, which requires two operands), it could ICE.

In `ConcretizeBoundsExprWithArgs` I added a `SubstitutedModifyingExpression` flag which would perform the same substitution, but it would also add a compiler error using Sema.Diag, so we knew compilation would abort before code generation.

**Comment by dtarditi** [comment]
This will be fixed PR #464.   We should return ExprError() instead of ExprResult().  This indicates that ExprResult is invalid.

**Comment by dtarditi** [comment]
This is fixed by https://github.com/Microsoft/checkedc-clang/pull/464.
